go/marshal/encode.go

```diff
@@ -0,0 +1,291 @@
+// Copyright 2016 Attic Labs, Inc. All rights reserved.
+// Licensed under the Apache License, version 2.0:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Package marshal implements encoding and decoding of Noms values. The mapping between JSON objects and Go values is described  in the documentation for the Marshal and Unmarshal functions.
+package marshal
+
+import (
+	"reflect"
+	"runtime"
+	"sort"
+	"strings"
+	"sync"
+	"unicode"
+
+	"github.com/attic-labs/noms/go/types"
+)
+
+// Marshal converts a Go value to a Noms value.
+//
+// Marshal traverses the value v recursively. Marshal uses the following type-dependent encodings:
+//
+// Boolean values are encoded as Noms types.Bool.
+//
+// Floating point and integer values are encoded as Noms types.Number. At the moment this might lead to some loss in precision because types.Number currently takes a float64.
+//
+// String values are encoded as Noms types.String.
+//
+// Struct values are encoded as Noms structs (types.Struct). Each exported struct field becomes a member of the struct unless
+//   - the field's tag is "-"
+// The structs default field name is the struct field name where the first character is lower cased,
+// but can be specified in the struct field's tag value. The "noms" key in
+// the struct field's tag value is the field name. Examples:
+//
+//   // Field is ignored.
+//   Field int `noms:"-"`
+//
+//   // Field appears in a Noms struct as field "myName".
+//   MyName int
+//
+//   // Field appears in a Noms struct as key "myName".
+//   Field int `json:"myName"`
+//
+// Unlike encoding/json Marshal, this does not support "omitempty".
+//
+// Anonymous struct fields are currently not supported.
+//
+// Noms values (values implementing types.Value) are copied over without any change.
+//
+// Go maps, slices, arrays, pointers, complex, interface (non types.Value), function are not supported. Attempting to encode such a value causes Marshal to return an UnsupportedTypeError.
```

Comment:
I see now that they're not, but it wasn't clear to me from this comment. I guess embedding structs leads to anonymous fields, so you do address it here, but I think it'd be good to call it out specifically.